### DISCUSSION TITLE
Toolbar that pops up when selecting text

### DISCUSSION THREAD

**Discussion by daxmonson** [chrisgurney/obsidian-note-toolbar#120] [Ideas]
Would be very interesting if when selecting text, a toolbar would pop up right above it. Similar to Popclip https://www.popclip.app/

**Comment by chrisgurney** [comment]
@daxmonson Thanks for the suggestion! I've been thinking about other places toolbars might make sense to show.

1. What sorts of tools would you want in a toolbar like this?
2. Would you want a toolbar for the note to still display (at the top, for example)?

**Comment by daxmonson** [comment]
@chrisgurney Sure, some quick thoughts:

1.
I'm thinking of features like copying text (or block links) to another note, or moving it to another note with other plugins like Text Transporter. Or, sending a selection to a header in a future daily note with the Review plugin. Inserting time stamps for keeping hours or for writing interstitial journals would be useful. Simple annotation options like italic, bold, highlights (or bracket links []) available there could reduce the cognitive load of going to toolbars further away from the text and then coming back to find your place. Tons of ideas with AI integration as well. Maybe the ability to trigger searches from selections to outside sites like arxiv.org, PubMed, translators, etc. would be helpful.

2.
Would want the ability to keep the top one there as well as one on selection. Can understand how that could get complicated quick code-wise. 

Some of these use cases make more sense on tablets without keyboards when doing heavy research and annotation. That can get tiresome even when going the minimal distance to toolbars elsewhere. Also, I wouldn't mind a vertical, always-expanded toolbar on the left or right margin (for tablet portrait reading and tools within thumb distance), but maybe that's already possible with the Style Settings plugin? I've pretty much made my own version of Readwise Reader in a vault with plugins and yours as been one of the most helpful for that. I'm just throwing some ideas out; I totally understand if it's out of scope! You've already done really excellent work.

**Comment by Moyf** [comment]
There is a plugin called PopKit https://github.com/zhouhua/obsidian-popkit

Maybe that's what you want 
![image](https://github.com/user-attachments/assets/8bc17e1d-0a78-4dd1-9949-13ad667e89cc)

**Comment by daxmonson** [comment]
@Moyf  thanks, that is similar, but not working on mobile here. As soon as it's engaged, general menu items (like even exiting plugin options) no longer are responsive to touch. Have to force close the app and reopen to disable it. Tried it on a vault with no other plugins and on the default theme and get the same which is weird. Dev can't reproduce it. Works fine on desktop though.

**Comment by FelipeRearden** [comment]
Hello @chrisgurney !!!

Just to add my humble `+1` for this feature suggested from @daxmonson 

About Moyf comment:

> There is a plugin called PopKit https://github.com/zhouhua/obsidian-popkit

Yes it is true, we have a plugin that could be used for this workflow. 

BUT there are 2 things that could make this feature inside NTB a 1000% powerful than PopKit

1️⃣ Using the already developed Folder Mapping  structure for the toolbars, we could have this `context ntb menu` mapped to specific files (via yaml), folders (map settings). Can you guys imagine how spectacular would be to select text and have a taylored set of items 🙏

2️⃣ All the benefits that are exclusive from NTB (File type, Menu type) could be expanded to this new tool. 

And the secret power is that @chrisgurney is an spectacular developer what can make this feature beautiful and consistent at the same time.

In a nutshell : +1 for this feature

**Comment by chrisgurney** [comment]
Thank you all for your thoughts. I have been thinking about toolbars for finer-grained contexts and how to accomplish the below. 

(None of these are off the table, at the moment.)

- Reading vs Editing view
- File types
- File patterns
- Tags: Think about a note that passes though stages, like `draft` → `to publish`, and having different toolbars to support those.
- Note headings
- Text actions

Ideas for both an easy-to-use settings/mappings UI, and also an intuitive toolbar UI point-of-view are much appreciated. Screenshots from other apps are welcome too. Keep those ideas coming!

**Reply by MrComputron** [reply]
Oh now that's a ** very** interesting idea. Context specific toolbars at a folder level is one thing,  but context specific to frontmatter property (draft vs publish) would allow a spectacular level of granularity.

I imagine if you are able to set the parameters determining the toolbar to a single field and value (eg. Status: Draft) you might be able to make it even more specific with multiple requirements (eg. Status: Draft AND Platform: Website), essentially creating a set of filters to determine the output toolbar.

This would ask the question of what do in cases of conflicting toolbar results arising from multiple filters.

**Reply by iiz00** [reply]
Hi, I’m a relatively new user of this plugin. Thank you for the amazing plugin!
I’d be really happy to see more flexible toolbar mappings in the future.

Regarding the mapping UI: one of my favorite Obsidian plugins is Iconic (https://github.com/gfxholo/iconic
), which assigns icons to notes based on configurable rules. Its UI for managing multiple rules (each with multiple conditions) is very clear and intuitive, and I think a similar approach could work well for Note Toolbar’s mapping interface. You might find it interesting to take a look.

I realize this is an older discussion, so apologies if this suggestion is no longer relevant. Wishing continued development and success for the plugin!

**Reply by chrisgurney** [reply]
@iiz00 Thanks for the suggestion about taking a look at Iconic's rules interface. In fact, I complimented @gfxholo when it was released. :)

Revamping Note Toolbar's rules, and adding more granularity, is the next major feature on my roadmap, so this is a timely suggestion/reminder!

**Comment by cinnabar777** [comment]
I've been messing about with a few plugins and other features working on making obsidian more mobile friendly — I only use Android devices and do all my markdown with my phone 😱

I mainly do articles published to WordPress and ebooks (old books out of copyright) along with the occasional notes and lists — school days are long in the rearview mirror 😂. That's too say I'm more of a normal word processor user than someone inter-linking notes and the like.

So everything I'm saying should be taken from a user, not a developer, exclusively using mobile 😁

Three toolbar plugins have come to be pretty useful for me.

1. [EDITING TOOLBAR](https://github.com/PKM-er/obsidian-editing-toolbar)
2. [POPKIT](https://github.com/zhouhua/obsidian-popkit)
3. [GAY TOOLBAR](https://github.com/ChasKane/gay-toolbar)

I'll explain what works for me on each and what an overall toolbar menu would look like, IMO 😁 maybe it'll be useful info coming from a non-dev.

Editing Toolbar:

The big positive here is submenus — pop-up secondary toolbars. This allows me to collect items by groups: text formatting, insert items, menu items, etc. 

Here's my setup; 

Main toolbar:

![editing-toolbar-01](https://github.com/user-attachments/assets/de11568d-ca7c-4013-a47f-1211eeee5a18)

Submenu for insert items: 

![editing-toolbar-02](https://github.com/user-attachments/assets/466a7e32-f8e2-451e-b004-064bbfa58c96)

Submenu for menu items:

![editing-toolbar-03](https://github.com/user-attachments/assets/f1c53469-0094-4461-9040-8eee88cffa9f)

Submenu for paragraph items:

![editing-toolbar-04](https://github.com/user-attachments/assets/0a8d2a40-d38c-453f-a5fc-f41bcad2a5c1)

Because editing toolbar plugin has a setting to fix the number of icons on the main toolbar I've not been able to set it to scroll — I have had to leave that as is — but I have been able to set the submenu to scroll therefore I can fit more than ten items while maintaining usability, can also alter width & height for bigger fingers — through css which isn't optimal for average users, if wide adoption is the goal.

The positives? Being able to have everything within one or two taps — this is more like a word processor. On mobile the command palette and slash command are rather slow compared to a quick tap. IMO developers are creating mobile apps from a desktop first perspective which leads to ideas like the command palette and slash commands instead of a fast access toolbar. 

Gay toolbar plugin: 

![gay-toolbar](https://github.com/user-attachments/assets/33d326dc-dee8-41a2-bd34-0082103ca27a)

This is my setup. Once again everything at the bottom (title & nav bars hidden). The bonus for GT is long press action — tap for primary action and long press for secondary allowing you to fit 20 actions into the standard ten icon row. The toolbar can do multiple rows and adjust the number of icons per row, adjust the icon size, and adjust the colors — very user friendly setup and clearly designed for mobile first.. This toolbar allows you to shove everything into a smaller space. Really in two rows you've got nearly everything quickly accessable, but of course I'm only using one row because I want to maximize screen.

POPKIT PLUGIN:

![POPKIT](https://github.com/user-attachments/assets/ddb115a6-9a2a-4b5f-a509-56c60050ffbc)

A simple floating toolbar when text is selected — WPS office started doing something similar a while back. IMO on mobile this is one of the most useful toolbars — very quick access to commonly used items. 

All three of the above toolbars are in early stages of development, but they show what a good combined toolbar, especially for mobile, would look like: 

1. A floating toolbar for selected text = fast actions. 
2. A bottom toolbar, that the thumb can reach easily, hiding unnecessary screen wasting bars (nav & title) maximizing the screen, designed for mobile first instead of desktop then hacked for mobile. 
3. Double actions — tap and long press
    1. Additionally adding gestures to increase quick actions: swipe up, down, left, and right for actions. Gay toolbar says they are working on this 🤷🏼‍♂️
4. Submenus for grouping items so everything is available within two taps or less — you don't have to go searching the command palette or slash commands except in rare situations.

That's a toolbar that would rival a good word processor and would have fast action for mobile users.

Customizable icon sizes are important on mobile — this is generally lacking and has to be done manually in css, not something average users are capable of. 

I understand that note toolbar plugin is about a customized toolbar per note or folder, etc — something I don't get 🤷🏼‍♂️ — I want everything available with the quickest access possible, but I'm certainly not a massive power user of obsidian. It's more like a lightweight word processor for me. Nonetheless, IMO, this type (what I'm suggesting) of toolbar becomes infinitely flexible under such conditions. 

The main parts are the basic functionality of a toolbar — being able to group things into submenus, multiple quick actions per icon, quick action floating menu — then it's simply specializing a setup for a particular use either for all files or a specific item(s). 

Now I've played a little with note toolbar, not that much because, from what I've seen it doesn't have: 

1\. Submenus that are quickly accessable. 

I see some quick action but that opens up a top command palette like thing which seems very cumbersome, comparatively. 

2\. No floating quick action toolbar. 

It has toolbars floating here and there, which isn't bad but the floating quick action should be like the system quick action menu — right there where your finger is. 

Therefore, while I can make up toolbars for specific notes or folders I can't make up a good comprehensive one, IMO. 

Just food for thought!

**Reply by chrisgurney** [reply]
@cinnabar777 Thank you for sharing your thoughts and for writing all this out! Definitely some good food for thought here!

**Comment by GitOnTop** [comment]
I also think this would be a very useful feature! There are issues with the existing text formatting plugins in Obsidian I've tried ("Editing Toolbar" doesn't close when/where it's supposed to, and cMenu doesn't have an option to "follow" the cursor). Popkit is a good idea, but has been a bit buggy (for me at least), and options for custom buttons are limited. What I like about Note Toolbar is there are a TON of configuration options. I've only scratched the surface of it so far, but it has been super stable for me. An option to have a 'text formatting' toolbar appear wherever text is highlighted would be sweet!

Regardless, big THANK YOU to the dev for all the work you've done with this -- it's truly a top notch plugin :)

**Reply by codey-badger** [reply]
About Editing Toolbar, on mobile, I can't have multiple configurations working at the same time.

**Comment by chrisgurney** [comment]
@GitOnTop @daxmonson @FelipeRearden @daxmonson @Moyf @MrComputron 

You can now select a toolbar that will appear when text is selected, in [`v1.26-beta-02`](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.26-beta-02).

If you give a try via [BRAT](https://github.com/TfTHacker/obsidian42-brat), let me know what you think!

---

**Mapping**

You can only select one toolbar that is used globally, for the moment. On the Roadmap: Mappings will get revamped, and as a result text toolbars will also get more flexible.

**Script support**

Below is an example of how to get the selected text via scripts. (I'm open to adding a simpler, one-line helper function for this, to the API.)


```
const e = app.workspace.activeLeaf.view?.editor;
const s = e.getSelection();
new Notice(s);
```

**Reply by Moyf** [reply]
Hey Chris, that's a surprising update!

I've tested it ;)

Some quick feedback:
1. It feels really good in general, and I love the concept.

2. Though I haven't figured out a specific condition, maybe we could **add a command/hotkey to show the "Text toolbar" directly at the cursor?** This way, even when no text is selected, we can suggest some operations based on the cursor position (like moving the entire line, etc.). Some commands can already be executed whether text is selected or not.

3. Looking forward to the mapping system! If notes in different folders could use different text toolbars, it would be very practical. 

Anyway, great job!

**Reply by chrisgurney** [reply]
> maybe we could add a command/hotkey to show the "Text toolbar" directly at the cursor?

I like this. I think I'll try having the `Focus on text toolbar` command make the toolbar appear at the cursor, if it's not already there.

**Reply by codey-badger** [reply]
I'm looking forward to how this develops. On mobile, when I need to move the whole line I usually do 3 taps but if the line has format or links then it doesn't get selected fully so it becomes frustrating to manually fix the selection or to start to search the "select line" command.

**Reply by chrisgurney** [reply]
@codey-badger 

> when I need to move the whole line...

Well this should make it easier then!

Add `Move line up` and `Move line down` commands (or from the [Note Toolbar Gallery](https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Gallery)) to a _Text toolbar_, and you should be good to go. No need to select the whole line; just drag the cursor to select a bit of text so the toolbar appears. 

Just tried it on my phone and it works great. The toolbar even stays available as you move the line up or down.

**Reply by chrisgurney** [reply]
@Moyf 

> maybe we could add a command/hotkey to show the "Text toolbar" directly at the cursor?

This is now implemented in [`v1.26-beta-04`](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.26-beta-04).

This is actually pretty useful as commands that work on cursor position (e.g., bold, italic) will bold/italicize the entire word.

Adding an API to get the word at the current position might be useful as well.

**Reply by Moyf** [reply]
<img width="504" height="105" alt="Snipaste_2025-11-11_11-50-16" src="https://github.com/user-attachments/assets/d6d51069-05eb-4e86-b961-a18c9b2c65b3" />
@chrisgurney Could you consider adding specific classes or data attributes to the "Text toolbar" element so we can modify it with CSS more easily? I understand we can currently use the "class for toolbar" to style it, but in the future there might be cases where:
1. The same toolbar is used in different locations, requiring specific styling **just for the text toolbar**
2. Multiple text toolbars exist, needing unified styling without adding classes to each one individually

**Reply by Moyf** [reply]
> Could you consider adding specific classes or data attributes to the "Text toolbar" element so we can modify it with CSS more easily?

Ah no worry! I think I missed it - `data-tbar-position="text"` is what I wanted, right?   :)

**Reply by chrisgurney** [reply]
@Moyf Style Settings support (CSS variables) is on my list for this release. 🙂

In the meantime you can use that data attribute, yes.

**Comment by daxmonson** [comment]
@chrisgurney  just installed and it's working great with everything I've put in so far. The only thing I've noticed is it's not opening in Hover Editor/Float Preview pop-ups. I've got a bunch of notes automated to open to headers that way in the day. Other toolbars are working fine in there so hope that wouldn't be too much to ask to get it working there?

Really appreciate you getting around to that original request and I can see this getting a ton of use. Thanks!

**Reply by chrisgurney** [reply]
Thank you @daxmonson !

> it's not opening in Hover Editor/Float Preview pop-ups

I think I can get this working. Will look at it this week.

**Reply by chrisgurney** [reply]
@daxmonson The toolbar should now appear overtop of popovers and Hover Editor windows, as of [`v1.26-beta-03`](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.26-beta-03).

Give it a try and let me know if it works for you!

**Comment by daxmonson** [comment]
@chrisgurney  It's working perfect on desktop in Hover Editor now! On mobile, there's something going on with style settings where when selecting text to get Note Toolbar to pop-up, the screen goes blank. This is with a regular note (or Hover). **It does work well on mobile with other plugins off,** so figured there was a conflict with another plugin.

# Steps 
 - Disabled all mobile plugins 
 - Flipped on Note Toolbar on mobile, works fine!
 - Flipped on Hover Editor, Note Toolbar is still working fine. 
 - Flipped on Style settings and got the screen going blank on selecting text.
 - Turn style settings off again and the toolbar works 
 
In the GIF, you can see with style settings on, there's a floating toolbar 3/4 of the way up on the left, possibly a conflict with that? I did try putting it places other than left, but that didn't help. The screen goes blank when trying to select the same text as is highlighted in the pic. Will probably have to get back to this fully on the weekend, but happy to help if you have any ideas on what it could be.

iOS 26.1, latest Obsidian Catalyst version 1.10.2 build 249
![IMG_3551](https://github.com/user-attachments/assets/4f153d35-4e69-46f8-9a0a-03beeef9332e)
![CleanShot 2025-11-10 at 19 42 15](https://github.com/user-attachments/assets/d40ad27b-b4ff-463a-8570-a472d3e0f2ad)

**Comment by daxmonson** [comment]
On a fresh test vault, it's working great on mobile with style settings so it's definitely a plugin conflict and more likely a style settings conflict. Will report back once I sort that. All good on your end!

**Reply by chrisgurney** [reply]
@daxmonson Is Hover Editor the issue here? Your steps included it but I haven't tried it yet on mobile/tablet.

**Comment by daxmonson** [comment]
@chrisgurney Ok, fixed it! This was the issue on my end in style settings: "note-toolbar@@cg-nt-tbar-margin-y": 55 
Somehow, I punched that in last time I was adjusting things instead of 0.5. To be clear, **working great on Desktop, mobile and in Hover Editor.** Thanks again!

**Comment by chrisgurney** [comment]
This feature is now available in [`v1.26.01`](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.26.01) via Community Plugins. Thanks all for your ideas and testing!

**Reply by MrComputron** [reply]
This is absolutely fantastic!
